sisyphus/job.py

```diff
@@ -651,9 +656,96 @@ def _sis_all_path_available(self):
                 return False
         return True
 
+    def _sis_get_file_stats(self) -> List[Tuple[str, float, int]]:
+        """
+        Returns a triple for every file below `work` and `output`: path, modification time, size.
+        
+        Path is relative to the job dir, the modification time is epoch time.
+
+        These stats are written to the `usage` files by the `LoggingThread`, and read by
+        Job._sis_get_expected_file_sizes.
+
+        """
+        stats = []
+        below_work = pathlib.Path(self._sis_path(gs.WORK_DIR)).rglob("*")
+        below_output = pathlib.Path(self._sis_path(gs.OUTPUT_DIR)).rglob("*")
+        for p in chain(below_work, below_output):
+            if p.is_file():
+                stat = p.stat()
+                rel_path = str(p.relative_to(self._sis_path()))
+                stats.append((rel_path, stat.st_mtime, stat.st_size))
+
+        return stats
+
+
+    @staticmethod
+    def _sis_get_expected_file_sizes(job: Union[str,"Job"], task: str = None,
+                                     timeout = gs.MAX_WAIT_FILE_SYNC) -> Dict[str, int]:
+        """
+        Tries to obtain the expected file sizes for files below `output` and `work` from the usage
+        files of the given job (or job dir). Returns None if the job had already been cleaned up.
+
+        If a usage file does not contain the file size information, this is either because the
+        respective task is still runnung or because the usage file is not yet synced. In this case,
+        retry until timeout and raise a TimeoutError.
+
+        When accumulating the information from several files, the most recent size info is retained
+        for every *existing* path. That is, deleted files are not part of the returned list.
+
+        The file paths returned are relative to the experiment directory.
+
+        If `task` is given, only usage files from these tasks are read.
+
+        """
+        # The job might be a Job object or the job directory
+        try:
+            job_dir = job._sis_path()
+        except AttributeError:
+            job_dir = job
+
+        if os.path.exists(os.path.join(job_dir, gs.JOB_FINISHED_ARCHIVE)):
```

Comment:
This could alternatively also be checked doing something like `subprocess.check_output(['tar', '-Oxf', os.path.join(job_dir, gs.JOB_FINISHED_ARCHIVE), 'usage.run.1'])`, but I'm not sure if this would be worth the effort. The given solution should work in nearly all cases if the clean up timeout is large enough.